editor/src/core/shared/multiplayer-hooks.tsx

```diff
@@ -37,3 +55,180 @@ export function useMyUserId(): string | null {
+    return []
+  }
+  return conns[myUserId] ?? []
+}
+
+/**
+ * Store the current user's connection in the room storage.
+ */
+export function useStoreConnection() {
+  const loginState = useEditorState(
+    Substores.userState,
+    (store) => store.userState.loginState,
+    'useAddConnection loginState',
+  )
+
+  const storeConnection = useMutation(
+    ({ storage, self }) => {
+      if (!isLoggedIn(loginState)) {
+        return
+      }
+
+      const selfConns: ConnectionInfo[] = storage.get('connections').get(self.id) ?? []
+
+      // if the current connection is not stored yet...
+      if (!selfConns.some((c) => c.id === self.connectionId)) {
+        // ...store it...
+        const now = Date.now()
+        selfConns.push({
+          id: self.connectionId,
+          startedAt: now,
+          lastSeen: now,
+        })
+        storage.get('connections').update({ [self.id]: selfConns })
+      }
+    },
+    [loginState],
+  )
+
+  const connections = useStorage((store) => store.connections)
+
+  React.useEffect(() => {
+    if (connections == null) {
+      return
+    }
+    storeConnection()
+  }, [connections, storeConnection])
+}
+
+/**
+ * Update the stored connection's heartbeat field and cleanup stale connections.
+ */
+export function useMonitorConnection() {
+  const debug = isFeatureEnabled('Debug – Connections')
+  const loginState = useEditorState(
+    Substores.userState,
+    (store) => store.userState.loginState,
+    'useMonitorConnection loginState',
+  )
+
+  const cleanupInactive = useMutation(
+    ({ storage }) => {
+      if (!isLoggedIn(loginState)) {
+        return
+      }
+
+      // get all conns by player ids
+      const allConns = storage.get('connections').toObject()
+
+      let changed = false
+      const now = Date.now()
+
+      // for each player id...
+      for (const playerId of Object.keys(allConns)) {
+        // iterate its connections...
+        allConns[playerId] = allConns[playerId].filter((conn) => {
+          // ...and if a connection is expired, remove it
+          const expired = now - conn.lastSeen > ConnectionExpiredTime
+          changed ||= expired
+          if (debug) {
+            console.info(`connection ${playerId}-${conn.id} expired=${expired}`)
+          }
+          return !expired
+        })
+      }
+
+      // if any connections were deleted, update the connections.
+      if (changed) {
+        if (debug) {
+          console.info('cleaning up expired connections')
+        }
+        storage.set('connections', new LiveObject(allConns))
+      }
+    },
+    [loginState],
+  )
+
+  const updateLastSeen = useMutation(
+    ({ storage, self }) => {
+      if (!isLoggedIn(loginState)) {
+        return
+      }
+
+      const selfConns: ConnectionInfo[] = storage.get('connections').get(self.id) ?? []
+      // if the current connection is stored, bump its lastSeen field
+      const connIndex = selfConns.findIndex((c) => c.id === self.connectionId)
```

Comment:
Doesn't look like you use `connIndex` as an index, so you could do:
```suggestion
      const doUpdate = selfConns.some((c) => c.id === self.connectionId)
```

Reply:
ahhh stupid remnant of a previous impl. thanks!